File: sasmodels/models/core_shell_bicelle_elliptical_protein.c
// NOTE that "length" here is the full height of the core!
static double
form_volume(double r_minor,
        double x_core,
        double thick_rim,
        double thick_face,
        double length,
        double radius_protein,
        double x_protein,
        double thick_protein)
{
    return M_PI*(  (r_minor + thick_rim)*(r_minor*x_core + thick_rim)* length +
                 square(r_minor)*x_core*2.0*thick_face +
                 square(radius_protein)*x_protein*2.0*thick_protein );
}

static double
radius_from_excluded_volume(double r_minor, double x_core, double thick_rim, double thick_face, double length,
 double radius_protein, double x_protein, double thick_protein)
{
    const double r_equiv     = sqrt((r_minor + thick_rim)*(r_minor*x_core + thick_rim));
    const double length_tot  = length + 2.0*thick_face +2.0*thick_protein;
    return 0.5*cbrt(0.75*r_equiv*(2.0*r_equiv*length_tot + (r_equiv + length_tot)*(M_PI*r_equiv + length_tot)));
}

static double
radius_from_volume(double r_minor, double x_core, double thick_rim, double thick_face, double length,
 double radius_protein, double x_protein, double thick_protein)
{
    const double volume_bicelle = form_volume(r_minor, x_core, thick_rim,thick_face,length, radius_protein,
    x_protein, thick_protein);
    return cbrt(volume_bicelle/M_4PI_3);
}

static double
radius_from_diagonal(double r_minor, double x_core, double thick_rim, double thick_face, double length,
 double radius_protein, double x_protein, double thick_protein)
 // RKH SORT THIS - do we want to allow for diagonal of protein ???
{
    const double radius_max = (x_core < 1.0 ? r_minor : x_core*r_minor);
    const double radius_max_tot = radius_max + thick_rim;
    const double length_tot = length + 2.0*thick_face;
    return sqrt(radius_max_tot*radius_max_tot + 0.25*length_tot*length_tot);
}

static double
radius_effective(int mode, double r_minor, double x_core, double thick_rim, double thick_face, double length,
 double radius_protein, double x_protein, double thick_protein)
{
    switch (mode) {
    default:
    case 1: // equivalent cylinder excluded volume
        return radius_from_excluded_volume(r_minor, x_core, thick_rim, thick_face,
                                length, radius_protein, x_protein, thick_protein);
    case 2: // equivalent sphere
        return radius_from_volume(r_minor, x_core, thick_rim, thick_face,
                                length, radius_protein, x_protein, thick_protein);
    case 3: // outer rim average radius
        return 0.5*r_minor*(1.0 + x_core) + thick_rim;
    case 4: // outer rim min radius
        return (x_core < 1.0 ? x_core*r_minor+thick_rim : r_minor+thick_rim);
    case 5: // outer max radius
        return (x_core > 1.0 ? x_core*r_minor+thick_rim : r_minor+thick_rim);
    case 6: // half outer thickness
        return 0.5*length + thick_face;
    case 7: // half diagonal (this ignores the missing "corners", so may give unexpected answer if thick_face
            //  or thick_rim is extremely large, ALSO ignores the protein)
        return radius_from_diagonal(r_minor,x_core,thick_rim,thick_face,
                                length, radius_protein, x_protein, thick_protein);
    }
}

static void
Fq(double q,
        double *F1,
        double *F2,
        double r_minor,
        double x_core,
        double thick_rim,
        double thick_face,
        double length,
        double rhoc,
        double rhoh,
        double rhor,
        double rhosolv,
        double sigma,
        double r_prot_minor,
        double x_prot,
        double thick_prot,
        double rhoprot)
{
      // core_shell_bicelle_elliptical_protein, RKH 24th April 2019, based on ..
     // core_shell_bicelle_elliptical_belt_rough, RKH 5th Oct 2017, based on core_shell_bicelle_elliptical
    const double halfheight = 0.5*length;
    const double r_major = r_minor * x_core;
    const double r2A = 0.5*(square(r_major) + square(r_minor));
    const double r2B = 0.5*(square(r_major) - square(r_minor));
    const double vol1 = M_PI*r_minor*r_major*(2.0*halfheight);
    const double vol2 = M_PI*(r_minor+thick_rim)*(r_major+thick_rim)*2.0*halfheight;
    const double vol3 = M_PI*r_minor*r_major*2.0*(halfheight+thick_face);
    // dr1,2,3 are now for Vcore, Vcore+rim, Vcore+face,
    const double dr1 = vol1*(-rhor - rhoh + rhoc + rhosolv);
    const double dr2 = vol2*(rhor-rhosolv);
    const double dr3 = vol3*(rhoh-rhosolv);
    // three more terms for the protein, which can stick out of the face (on both sides simultaneously)
    const double r_prot_major = r_prot_minor * x_prot;
    const double r2A_prot = 0.5*(square(r_prot_major) + square(r_prot_minor));
    const double r2B_prot = 0.5*(square(r_prot_major) - square(r_prot_minor));
    const double area_prot = M_PI*r_prot_minor*r_prot_major;
    const double dr4 = area_prot*2.0*(halfheight + thick_face + thick_prot)*(rhoprot - rhosolv);
    // note this next term really does need (-rhoh +rhosolv)
    const double dr5 = area_prot*2.0*(halfheight + thick_face)*(rhosolv -rhoh);
    const double dr6 = area_prot*2.0*halfheight*(rhoh-rhoc);

    //initialize integral
    double outer_total_F1 = 0.0;
    double outer_total_F2 = 0.0;
    for(int i=0;i<GAUSS_N;i++) {
        //setup inner integral over the ellipsoidal cross-section
        // since we generate these lots of times, why not store them somewhere?
        //const double cos_theta = ( GAUSS_Z[i]*(vb-va) + va + vb )/2.0;
        const double cos_theta = ( GAUSS_Z[i] + 1.0 )/2.0;
        const double sin_theta = sqrt(1.0 - cos_theta*cos_theta);
        const double qab = q*sin_theta;
        const double qc = q*cos_theta;
        const double si1 = sas_sinx_x(halfheight*qc);
        const double si2 = sas_sinx_x((halfheight + thick_face)*qc);
        const double si3 = sas_sinx_x((halfheight + thick_face + thick_prot)*qc);
        double inner_total_F1 = 0;
        double inner_total_F2 = 0;
        for(int j=0;j<GAUSS_N;j++) {
            //76 gauss points for the inner integral (WAS 20 points,so this may make unnecessarily slow, but playing safe)
            //const double beta = ( GAUSS_Z[j]*(vbj-vaj) + vaj + vbj )/2.0;
            const double beta = ( GAUSS_Z[j] +1.0)*M_PI_2;
            const double rr = sqrt(r2A - r2B*cos(beta));
            const double be1 = sas_2J1x_x(rr*qab);
            const double be2 = sas_2J1x_x((rr+thick_rim)*qab);
            const double rrp = sqrt(r2A_prot - r2B_prot*cos(beta));
            const double be4 = sas_2J1x_x(rrp*qab);
            const double f = dr1*si1*be1 + dr2*si1*be2 + dr3*si2*be1 +
                             be4*( dr4*si3 + dr5*si2 +dr6*si1);

            inner_total_F1 += GAUSS_W[j] * f;
            inner_total_F2 += GAUSS_W[j] * f * f;
        }
        //now calculate outer integral
        outer_total_F1 += GAUSS_W[i] * inner_total_F1;
        outer_total_F2 += GAUSS_W[i] * inner_total_F2;
    }
    // now complete change of integration variables (1-0)/(1-(-1))= 0.5
    outer_total_F1 *= 0.25;
    outer_total_F2 *= 0.25;

    //convert from [1e-12 A-1] to [cm-1]
    *F1 = 1e-2*outer_total_F1*exp(-0.25*square(q*sigma));
    *F2 = 1e-4*outer_total_F2*exp(-0.5*square(q*sigma));
}

static double
Iqabc(double qa, double qb, double qc,
          double r_minor,
          double x_core,
          double thick_rim,
          double thick_face,
          double length,
          double rhoc,
          double rhoh,
          double rhor,
          double rhosolv,
          double sigma,
          double r_prot_minor,
          double x_prot,
          double thick_prot,
          double rhoprot)
{
    // Vol1,2,3 and dr1,2,3 are now for Vcore, Vcore+rim, Vcore+face,
    const double dr1 = -rhor - rhoh + rhoc + rhosolv;
    const double dr2 = rhor-rhosolv;
    const double dr3 = rhoh-rhosolv;
    const double r_major = r_minor*x_core;
    const double halfheight = 0.5*length;
    const double vol1 = M_PI*r_minor*r_major*length;
    const double vol2 = M_PI*(r_minor+thick_rim)*(r_major+thick_rim)*2.0*halfheight;
    const double vol3 = M_PI*r_minor*r_major*2.0*(halfheight+thick_face);
    // three more terms for the protein
    const double r_prot_major = r_prot_minor * x_prot;
    const double area_prot = M_PI*r_prot_minor*r_prot_major;
    const double dr4 = area_prot*2.0*(halfheight + thick_face + thick_prot)*(rhoprot - rhosolv);
    // note this next term really does need (-rhoh +rhosolv)
    const double dr5 = area_prot*2.0*(halfheight + thick_face)*(rhosolv -rhoh);
    const double dr6 = area_prot*2.0*halfheight*(rhoh-rhoc);

    // Compute effective radius in rotated coordinates
    const double qr_hat = sqrt(square(r_major*qb) + square(r_minor*qa));
    // does this need to be changed for the "missing corners" where there there is no "belt" ?
    const double qrshell_hat = sqrt(square((r_major+thick_rim)*qb)
                                   + square((r_minor+thick_rim)*qa));
    const double qr_prot_hat = sqrt(square(r_prot_major*qb) + square(r_prot_minor*qa));
    const double be1 = sas_2J1x_x( qr_hat );
    const double be2 = sas_2J1x_x( qrshell_hat );
    const double be4 = sas_2J1x_x( qr_prot_hat );
    const double si1 = sas_sinx_x( halfheight*qc );
    const double si2 = sas_sinx_x( (halfheight + thick_face)*qc );
    const double si3 = sas_sinx_x( (halfheight + thick_face + thick_prot)*qc );
    const double fq = vol1*dr1*si1*be1 + vol2*dr2*si1*be2 +  vol3*dr3*si2*be1 +
                           be4*( dr4*si3 + dr5*si2 +dr6*si1) ;
    const double atten = exp(-0.5*(qa*qa + qb*qb + qc*qc)*(sigma*sigma));
    return 1.0e-4 * fq*fq * atten;
}

File: sasmodels/models/core_shell_bicelle_elliptical_protein.py
r"""
Definition
----------

Adds an elliptical rod shaped "protein" through the center of 
core_shell_bicelle_elliptical_belt_rough
The protein sticks out beyond the face of the bicelle, on BOTH faces, 
by distance thick_protein. The whole particle remains centrosymmetric.
The assymetric case, where the protein sticks out further on one face 
than the other is not yet possible in sasview,as that will require a 
completely different I(Q) calculation involving complex algebra.

NOTE it is up to the user to keep the elliptical protein cross section
inside the elliptical core of the bicelle, else non-physical results will
be produced!


.. figure:: img/core_shell_bicelle_elliptical_protein.png

   Schematic cross section of core_shell_bicelle_elliptical_protein. Users
   will have to decide how to distribute "heads" and "tails" between the
   rim, face and core regions in order to estimate appropriate starting
   parameters.

The scattering length densities (sld) are: $\rho_c$, the core, $\rho_f$,
the face, $\rho_r$, the rim, $\rho_p$, the protein,  and $\rho_s$ the solvent.

The form factor for the protein containing bicelle is calculated in cylindrical
coordinates, where
$\alpha$ is the angle between the $Q$ vector and the cylinder axis, and $\psi$
is the angle for the ellipsoidal cross section core, to give:

.. math::

    I(Q,\alpha,\psi) = \frac{\text{scale}}{V_t}
        \cdot F(Q,\alpha, \psi)^2 \cdot \sin(\alpha)
        \cdot\exp\left\{ -\frac{1}{2}Q^2\sigma^2 \right\} + \text{background}

where a numerical integration of $F(Q,\alpha, \psi)^2\sin(\alpha)$ is
carried out over $\alpha$ and $\psi$ for:

.. math::

    F(Q,\alpha,\psi) = &\bigg[
      (\rho_c -\rho_r - \rho_f + \rho_s) V_c
      \frac{2J_1(QR'\sin \alpha)}{QR'\sin\alpha}
      \frac{\sin(Q(L/2)\cos\alpha)}{Q(L/2)\cos\alpha} \\
    &+(\rho_f - \rho_s) V_{c+f}
      \frac{2J_1(QR'\sin\alpha)}{QR'\sin\alpha}
      \frac{\sin(Q(L/2+t_f)\cos\alpha)}{Q(L/2+t_f)\cos\alpha} \\
    &+(\rho_r - \rho_s) V_{c+r}
      \frac{2J_1(Q(R'+t_r)\sin\alpha)}{Q(R'+t_r)\sin\alpha} 
      \frac{\sin(Q(L/2)\cos\alpha)}{Q(L/2)\cos\alpha} \\
    &+\bigg((\rho_p - \rho_s) V_p\frac{\sin(Q(L/2 +t_f +t_p)\cos\alpha)}{Q(L/2 +t_f +t_p)\cos\alpha} \\
    &+(\rho_s - \rho_f) V_d\frac{\sin(Q(L/2 +t_f)\cos\alpha)}{Q(L/2 +t_f)\cos\alpha} \\
    &+(\rho_f - \rho_c) V_e\frac{\sin(Q(L/2)\cos\alpha)}{Q(L/2)\cos\alpha} \bigg)
    \frac{2J_1(QR'_p\sin\alpha)}{QR'_p\sin\alpha}\bigg]

where

.. math::

    R' = \frac{R}{\sqrt{2}}
        \sqrt{(1+X_\text{core}^{2}) + (1-X_\text{core}^{2})\cos(\psi)} \\
    R'_p = \frac{R_p}{\sqrt{2}}
        \sqrt{(1+X_\text{protein}^{2}) + (1-X_\text{protein}^{2})\cos(\psi)}


and $V_t = \pi (R+t_r)(X_\text{core} R+t_r) L + 2 \pi X_\text{core} R^2 t_f+2\pi X_\text{protein}R_p^2 thick_\text{protein}$ is
the total volume of the bicelle plus any protein sticking out from the faces,
$V_c = \pi X_\text{core} R^2 L$ the volume of
the core, $V_{c+f} = \pi X_\text{core} R^2 (L+2 t_f)$ the volume of the core
plus the volume of the faces, $V_{c+r} = \pi (R+t_r)(X_\text{core} R+t_r) L$
the volume of the core plus the rim, $V_d = \pi X_\text{protein} R_p^2 (L+2t_f)$ is the
volume of protein intersecting the bicelle, $V_e = \pi X_\text{protein} R_p^2 L$ is the
volume of protein intersecting the core of the bicelle.

$R$ is the radius of the core,
$X_\text{core}$ is the axial ratio of the core, $L$ the length of the core,
$t_f$ the thickness of the face, $t_r$ the thickness of the rim and $J_1$ the
usual first order bessel function. The bicelle core has radii $R$ and $X_\text{core} R$
so is circular, as for the core_shell_bicelle model, for $X_\text{core}=1$.
Note that you may need to limit the range of $X_\text{core}$, especially if
using the Monte-Carlo algorithm, as setting radius to $R/X_\text{core}$ and
axial ratio to $1/X_\text{core}$ gives an equivalent solution!

$R_p$ is the radius of the protein,
$X_\text{protein}$ is the axial ratio of the protein, which has total length $L+2t_f+2thick_\text{protein}$.
The axes of the elliptical cross section protein cylinder are common with the
bicelle, so making $X_\text{protein} < 1$ whilst $X_\text{core} > 1$ can effectively
rotate the protein by 90 degrees relative to the bicelle.

An approximation for the effects of "Gaussian interfacial roughness" $\sigma$
is included, by multiplying $I(Q)$ by
$\exp\left \{ -\frac{1}{2}Q^2\sigma^2 \right \}$. This applies, in some way, to
all interfaces in the model not just the external ones. It will only work well 
when $\sigma$ is very small compared to the particle size.(Note that for a one
dimensional system convolution of the scattering length density profile with
a Gaussian of standard deviation $\sigma$ does exactly this multiplication.)
Leave $\sigma$ set to zero for the usual sharp interfaces.

The output of the 1D scattering intensity function for randomly oriented
bicelles is then given by integrating over all possible $\alpha$ and $\psi$.

For oriented bicelles the *theta*, *phi* and *psi* orientation parameters
will appear when fitting 2D data, for further details of the calculation
and angular dispersions  see :ref:`orientation` .


.. figure:: img/elliptical_cylinder_angle_definition.png


    Definition of the angles for the oriented core_shell_bicelle_elliptical
    particles.



References
----------

.. [#] Solution structure of bilayer membrane-embedded proton-translocating 
pyrophosphatase revealed via small-angle X-ray scattering”, Orion Shih, Yi-Qi 
Yeh, Kuei-Fen Liao, Kun-Mou Li, Jia-Yin Tsai, Chieh-Chin Li, Yun-Wei Chiang, 
Richard K. Heenan, Yuh-Ju Sun & U-Ser Jeng. Materials Chemistry and Physics, 
308(2023)128253.  https://doi.org/10.1016/j.matchemphys.2023.128253

Source
------

Authorship and Verification
----------------------------

* **Author:** Richard Heenan **Date:** April 24, 2019
* **Last Modified by:Richard Heenan, Sept 2023**
* **Last Reviewed by:**
"""

from numpy import inf, sin, cos, pi

name = "core_shell_bicelle_elliptical_protein"
title = "Elliptical cylinder with a core-shell and protein through the center"
description = """
    core_shell_bicelle_elliptical_protein
    Elliptical cylinder core, optional shell on the two flat faces, and "belt" shell of
    uniform thickness on its rim (in this case NOT extending around the end faces),
    with approximate interfacial roughness, all with an elliptical cylinder of protein
    inserted through the center of the bicelle.
    Please see full documentation for equations and further details.
    Involves a double numerical integral around the ellipsoid diameter
    and the angle of the cylinder axis to Q.
    Compare also the core_shell_bicelle and elliptical_cylinder models.
      """
category = "shape:cylinder"

# pylint: disable=bad-whitespace, line-too-long
#             ["name", "units", default, [lower, upper], "type", "description"],
parameters = [
    ["radius",         "Ang",       30, [0, inf],    "volume",      "Cylinder core radius r_minor"],
    ["x_core",        "None",       3,  [0, inf],    "volume",      "Axial ratio of core, X = r_major/r_minor"],
    ["thick_rim",     "Ang",         8, [0, inf],    "volume",      "Rim or belt shell thickness"],
    ["thick_face",    "Ang",        14, [0, inf],    "volume",      "Cylinder face thickness"],
    ["length",         "Ang",       50, [0, inf],    "volume",      "Cylinder length"],
    ["sld_core",       "1e-6/Ang^2", 4, [-inf, inf], "sld",         "Cylinder core scattering length density"],
    ["sld_face",       "1e-6/Ang^2", 7, [-inf, inf], "sld",         "Cylinder face scattering length density"],
    ["sld_rim",        "1e-6/Ang^2", 1, [-inf, inf], "sld",         "Cylinder rim scattering length density"],
    ["sld_solvent",    "1e-6/Ang^2", 6, [-inf, inf], "sld",         "Solvent scattering length density"],
    ["sigma",          "Ang",     0,    [0, inf],    "",            "Interfacial roughness"],
    ["radius_protein", "Ang",       18, [0, inf],    "volume",      "Protein radius r_minor"],
    ["x_protein",     "None",     1.5,  [0, inf],    "volume",      "Axial ratio of protein, X = r_major/r_minor"],
    ["thick_protein",  "Ang",       12, [0, inf],    "volume",      "Extension of protein beyond faces"],
    ["sld_protein",    "1e-6/Ang^2", 2, [-inf, inf], "sld",         "Protein scattering length density"],
    ["theta",       "degrees",    90.0, [-360, 360], "orientation", "Cylinder axis to beam angle"],
    ["phi",         "degrees",    0,    [-360, 360], "orientation", "Rotation about beam"],
    ["psi",         "degrees",    0,    [-360, 360], "orientation", "Rotation about cylinder axis"],
    ]

# pylint: enable=bad-whitespace, line-too-long

source = ["lib/sas_Si.c", "lib/polevl.c", "lib/sas_J1.c", "lib/gauss76.c",
          "core_shell_bicelle_elliptical_protein.c"]
have_Fq = True
radius_effective_modes = [
    "equivalent cylinder excluded volume", "equivalent volume sphere",
    "outer rim average radius", "outer rim min radius",
    "outer max radius", "half outer thickness", "half diagonal",
    ]

# TODO: No random() for core-shell bicelle elliptical belt rough

demo = dict(scale=1, background=0,
            radius=30.0,
            x_core=3.0,
            thick_rim=8.0,
            thick_face=14.0,
            length=50.0,
            sld_core=4.0,
            sld_face=7.0,
            sld_rim=1.0,
            sld_solvent=6.0,
            radius_protein=18.0,
            x_protein=1.5,
            thick_protein=12.0,
            sld_protein=2.0,
            theta=90,
            phi=0,
            psi=0,
            sigma=0)

q = 0.1
# april 6 2017, rkh added a 2d unit test, NOT READY YET pull #890 branch assume correct!
qx = q*cos(pi/6.0)
qy = q*sin(pi/6.0)

tests = [
    #[{'radius': 30.0, 'x_core': 3.0, 'thick_rim':8.0, 'thick_face':14.0, 'length':50.0}, 'ER', 1],
    #[{'radius': 30.0, 'x_core': 3.0, 'thick_rim':8.0, 'thick_face':14.0, 'length':50.0}, 'VR', 1],
#
     # this test as per the original core_shell_bicelle_elliptical_belt_rough
    [{'radius': 30.0, 'x_core': 3.0, 'thick_rim': 8.0, 'thick_face': 14.0,
      'length': 50.0, 'sld_core': 4.0, 'sld_face': 7.0, 'sld_rim': 1.0,
      'radius_protein':0.0, 'x_protein':1.5, 'thick_protein':0.0, 'sld_protein':2.0,
      'sld_solvent': 6.0, 'background': 0.0},
     0.015, 189.328],
    #[{'theta':80., 'phi':10.}, (qx, qy), 7.88866563001 ],
]

del qx, qy  # not necessary to delete, but cleaner
